File: lib/entities/entities_helper.dart
import 'package:didomi_sdk/entities/purpose.dart';
import 'package:didomi_sdk/entities/vendor.dart';

import 'user_status.dart';

/// Utility class used to do computations related to vendors and purposes.
class EntitiesHelper {
  /// Convert a list of dictionaries into a list of purposes.
  static List<Purpose> toPurposes(List<dynamic> rawList) => rawList.map((e) => Purpose.fromJson(e)).toList();

  /// Convert a list of dictionaries into a list of vendors.
  static List<Vendor> toVendors(List<dynamic> rawList) => rawList.map((e) => Vendor.fromJson(e)).toList();

  /// Convert json to status Ids, or null
  static Ids? toIds(dynamic json) => json != null ? Ids.fromJson(json) : null;

  /// Convert json to purposes status info, or null
  static Purposes? toPurposesStatus(dynamic json) => json != null ? Purposes.fromJson(json) : null;

  /// Convert json to vendors status info, or null
  static Vendors? toVendorsStatus(dynamic json) => json != null ? Vendors.fromJson(json) : null;

}

File: ios/didomi_sdk/Classes/DidomiSwift/EntitiesHelper.swift
//
//  EntitiesHelper.swift
//  didomi_sdk
//
//  Created by Felipe Saez V on 1/06/21.
//

import Foundation
import Didomi

// Class used to do computations related to vendors and purposes.
class EntitiesHelper {
    // Convert an array of purposes into an array of dictionaries.
    static func dictionaries(from purposes: [Purpose]) -> [[String: Any]?] {
        return purposes.map { dictionary(from: $0) }
    }
    
    // Convert an instance of Purpose into a dictionary.
    static func dictionary(from purpose: Purpose?) -> [String: Any]? {
        let encoder = JSONEncoder()
        let data = try? encoder.encode(purpose)
        return dictionary(from: data)
    }
    
    // Convert an array of vendors into an array of dictionaries.
    static func dictionaries(from vendors: [Vendor]) -> [[String: Any]?] {
        return vendors.map { dictionary(from: $0) }
    }
    
    // Convert an instance of Vendor into a dictionary.
    static func dictionary(from vendor: Vendor?) -> [String: Any]? {
        let encoder = JSONEncoder()
        let data = try? encoder.encode(vendor)
        return dictionary(from: data)
    }
    
    // Convert an instance of CurrentUserStatus into a dictionary.
    static func dictionary(from currentUserStatus: CurrentUserStatus?) -> [String: Any]? {
        let encoder = JSONEncoder()
        let data = try? encoder.encode(currentUserStatus)
        return dictionary(from: data)
    }
    
    // Convert an instance of VendorStatus into a dictionary.
    static func dictionary(from vendorStatus: CurrentUserStatus.VendorStatus?) -> [String: Any]? {
        let encoder = JSONEncoder()
        let data = try? encoder.encode(vendorStatus)
        return dictionary(from: data)
    }

    // Convert an instance of UserStatus into a dictionary.
    static func dictionary(from userStatus: UserStatus?) -> [String: Any]? {
        let encoder = JSONEncoder()
        let data = try? encoder.encode(userStatus)
        return dictionary(from: data)
    }
    
    // Convert an instance of Data into a dictionary.
    static private func dictionary(from data: Data?) -> [String: Any]? {
        if let data = data {
            return try? JSONSerialization.jsonObject(with: data, options: .allowFragments) as? [String: Any]
        } else {
            return nil
        }
    }
}
